### ISSUE TITLE
Possible custom Request and Reply objects?

### ISSUE THREAD

**Issue by jsumners** [fastify/fastify#327]
There's a PR to Node under review that would add the ability to create an instance of `http.Server` with custom Request and Reply objects -- https://github.com/nodejs/node/pull/15752. If this gets accepted, I think it will be a powerful feature that Fastify could leverage. I'm opening this issue to start the discussion about such a possibility.

**Comment by allevo** [comment]
Wow! thanks for sharing the link!

Anyway that PR don't completely cover the fastify encapsulation feature. But if the context is the global maybe we can use it! We need to figure out how to support it with different node versions.

**Comment by mcollina** [comment]
It is going to get accepted, but it will take a while before we can use that everywhere. We should have a ponyfill situation.

**Comment by jsumners** [comment]
Update: this might be merged within the next two days from the time of this comment.

https://github.com/nodejs/node/pull/15752#issuecomment-361794346

**Comment by mcollina** [comment]
This is meat for a future release before it's available in all of our supported Node.js LTS lines.
It might simplify some of our internals, or maybe we can reengineer some code to actually leverage that.

**Comment by jsumners** [comment]
Right. I was just providing an update to keep people informed. This is in no way a blocker for v1.0.0

**Comment by mcollina** [comment]
I don't think we should really target this for v2, as we can't really backport it to Node 6.

**Comment by delvedor** [comment]
@mcollina Roadmap for `v3.0.0`?

**Comment by Ethan-Arrowood** [comment]
Fwiw I believe that having defined, custom Request and Reply objects will make the type declarations simpler.

**Comment by delvedor** [comment]
I've studied a little how this would work, and I think that for now, we should not pursue this path for two main reasons.
1. It's intended to work only for http 1.1, not sure what will happen for http 2 and http 3.
1. It's not compatible with our encapsulation model, as the encapsulation happens after the routing *(so the core request and response object are already initialized)* and we create our own after the routing.

I think we should come back on this once http 3 is settled, and we'll see how the API will evolve.
